Paintroid/src/main/java/org/catrobat/paintroid/tools/implementation/BrushTool.kt

```diff
@@ -0,0 +1,158 @@
+    toolOptionsViewController: ToolOptionsVisibilityController,
+    toolPaint: ToolPaint,
+    workspace: Workspace,
+    commandManager: CommandManager
+) : BaseTool(contextCallback, toolOptionsViewController, toolPaint, workspace, commandManager) {
+    protected open val previewPaint: Paint
+        get() = toolPaint.previewPaint
+
+    protected open val bitmapPaint: Paint
+        get() = toolPaint.paint
+
+    override val toolType: ToolType
+        get() = ToolType.BRUSH
+
+    @VisibleForTesting
+    lateinit var pathToDraw: Path
+    private var initialEventCoordinate: PointF? = null
+    private var pathInsideBitmap = false
+    private val drawToolMovedDistance = PointF(0f, 0f)
+
+    init {
+        pathToDraw = Path()
+        pathToDraw.incReserve(1)
+        brushToolOptionsView.setBrushChangedListener(CommonBrushChangedListener(this))
+        brushToolOptionsView.setBrushPreviewListener(CommonBrushPreviewListener(toolPaint, toolType))
+        brushToolOptionsView.setCurrentPaint(toolPaint.paint)
+    }
+
+    override fun draw(canvas: Canvas) {
+        canvas.run {
+            save()
+            clipRect(0, 0, workspace.width, workspace.height)
+            drawPath(pathToDraw, previewPaint)
+            restore()
+        }
+    }
+
+    override fun handleDown(coordinate: PointF?): Boolean {
+        coordinate ?: return false
+        initialEventCoordinate = PointF(coordinate.x, coordinate.y)
+        previousEventCoordinate = PointF(coordinate.x, coordinate.y)
+        pathToDraw.moveTo(coordinate.x, coordinate.y)
+        drawToolMovedDistance[0f] = 0f
+        pathInsideBitmap = workspace.contains(coordinate)
+        return true
+    }
+
+    override fun handleMove(coordinate: PointF?): Boolean {
+        if (initialEventCoordinate == null || previousEventCoordinate == null || coordinate == null) {
+            return false
+        }
+        previousEventCoordinate?.let {
+            pathToDraw.quadTo(it.x, it.y, coordinate.x, coordinate.y)
+            pathToDraw.incReserve(1)
+            drawToolMovedDistance.set(
+                drawToolMovedDistance.x + Math.abs(coordinate.x - it.x),
+                drawToolMovedDistance.y + Math.abs(coordinate.y - it.y)
+            )
+            it.set(coordinate.x, coordinate.y)
+        }
+        if (!pathInsideBitmap && workspace.contains(coordinate)) {
+            pathInsideBitmap = true
+        }
+        return true
+    }
+
+    override fun handleUp(coordinate: PointF?): Boolean {
+        if (initialEventCoordinate == null || previousEventCoordinate == null || coordinate == null) {
+            return false
+        }
+        previousEventCoordinate?.let {
+            drawToolMovedDistance.set(
+                drawToolMovedDistance.x + Math.abs(coordinate.x - it.x),
+                drawToolMovedDistance.y + Math.abs(coordinate.y - it.y)
+            )
+        }
+        if (!pathInsideBitmap && workspace.contains(coordinate)) {
```

Comment:
in Java this if statement was before the drawToolMoveDistance.set(..,..), what was the for reason for changing the order?

Reply:
oh.. that wasn't intentional 😅, but I think it doesn't make any difference either🤔